object_store/src/buffered.rs

```diff
@@ -400,6 +400,125 @@ impl AsyncWrite for BufWriter {
     }
 }
 
+/// A buffered multipart uploader.
+///
+/// This uploader adaptively uses [`ObjectStore::put`] or
+/// [`ObjectStore::put_multipart`] depending on the amount of data that has
+/// been written.
+///
+/// Up to `capacity` bytes will be buffered in memory, and flushed on shutdown
+/// using [`ObjectStore::put`]. If `capacity` is exceeded, data will instead be
+/// streamed using [`ObjectStore::put_multipart`]
+///
+/// # TODO
+///
+/// Add attributes and tags support.
+pub struct BufUploader {
+    store: Arc<dyn ObjectStore>,
+    path: Path,
+
+    chunk_size: usize,
+    max_concurrency: usize,
+
+    buffer: PutPayloadMut,
+    write_multipart: Option<WriteMultipart>,
+}
+
+impl std::fmt::Debug for BufUploader {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("BufUploader")
+            .field("chunk_size", &self.chunk_size)
+            .finish()
+    }
+}
+
+impl BufUploader {
+    /// Create a new [`BufUploader`] from the provided [`ObjectStore`] and [`Path`]
+    pub fn new(store: Arc<dyn ObjectStore>, path: Path) -> Self {
+        Self::with_chunk_size(store, path, 5 * 1024 * 1024)
+    }
+
+    /// Create a new [`BufUploader`] from the provided [`ObjectStore`], [`Path`] and `capacity`
```

Comment:
nit:

```suggestion
    /// Create a new [`BufUploader`] from the provided [`ObjectStore`], [`Path`] and `chunk_size`
```

Also, it seems strange to me this function is called `with_chunk_size` when it is creating a new instance. Perhaps calling it `new_with_chunk_size` would be more consistent

Reply:
I also don't understand why chunk_size is set via a special constructor(`with_chunk_size`) but `max_concurrency` is set with a builder style API to update an existing object (`with_max_concurrency` )

Perhaps we can use builder style API for both

```rust
    /// Specify the upload chunk size
    ///
    /// Defaults to 1MB
    pub fn with_chunk_size(self, chunk_size: usize) -> Self {
        Self {
            chunk_size,
            ..self
        }
    }
```

If there is a reason for the inconsistency, could you please add a comment explaining the rationale? (e.g. maybe changing chunk size after creation is not valid 🤔 )

Reply:
I try to keep the consistent API with existing `WriteMultipart`:

https://github.com/apache/arrow-rs/blob/087f34b70e97ee85e1a54b3c45c5ed814f500b0a/object_store/src/upload.rs#L121-L129

And yes, we can't change the chunk size after created a `BufUploader`. I will add a comment here. Or maybe we can make `chunk_size` a required field for `BufUploader::new`?